### ISSUE TITLE
Cannot access [Model] before initialization

### ISSUE THREAD

**Issue by alsonkemp** [RobinBuschmann/sequelize-typescript#825]
**Versions**

- sequelize: 4.44.4
- sequelize-typescript: 1.1.0
- typescript: 3.9.2

**I'm submitting a ...** 

[X] bug report
[ ] feature request

**Actual behavior:**

$ ./node_modules/.bin/ts-node test.ts
ReferenceError: Cannot access 'Team' before initialization

**Expected behavior:**

No error.

**Steps to reproduce:**
Use (basically) the code from the README as `test.ts`:
```ts
import {BelongsTo, Column, ForeignKey, HasMany, Model, Table} from "sequelize-typescript";

@Table
export class Player extends Model<Player> {

  @Column
  name!: string;

  @Column
  num!: number;

  @ForeignKey(() => Team)
  @Column
  teamId!: number;

  @BelongsTo(() => Team)
  team!: Team;
}

@Table
export class Team extends Model<Team> {

  @Column
  name!: string;

  @HasMany(() => Player)
  players!: Player[];
}
```
tsconfig.json:
```ts
{
  "compilerOptions": {
    "emitDecoratorMetadata": true,
    "esModuleInterop": true,
    "experimentalDecorators": true,
    "forceConsistentCasingInFileNames": true,
    "lib": ["es2019", "es2020.promise", "es2020.bigint", "es2020.string"],
    "noImplicitAny": true,
    "module": "commonjs",
    "outDir": "_dist",
    "rootDir": ".",
    "skipLibCheck": true,
    "sourceMap": false,
    "strict": true,
    "strictNullChecks": true,
    "target": "ES2019"
  }
}
```

**Comment by alsonkemp** [comment]
Switching to Sequelize({config: {models: ... } }) seems to work around this but I'm not sure why/how?

**Comment by yassernascimento** [comment]
I resolved that just putting each model in different files. It is not a sequelize issue, seems like the problem is on typescript compilation with `emitDecoratorMetadata: true` option.

**Comment by mv29** [comment]
I am also facing the same issue 

- Error = ReferenceError: Cannot access 'Conversation' before initialization
- sequelize version 6.50

I Have a contact model that has a belongsTo association with a Conversation Model. 

```
import sequelize from "sequelize";
import database from "../config/database.js";
import Conversation from "./conversation.js"

const { DataTypes } = sequelize;

const Contact = database.define("contact", {
  opted_in: {
    default: false,
    type: DataTypes.BOOLEAN,
  },
  uuid: {
    type: DataTypes.STRING,
  },
  country_code: {
    type: DataTypes.STRING,
    allowNull: false,
  },
  mobile_number: {
    type: DataTypes.STRING,
    allowNull: false,
  },
  name: {
    type: DataTypes.STRING,
  },
  email: {
    type: DataTypes.STRING,
    isEmail: true,
  },
});

Contact.hasOne(models.Conversation);

export default Contact;
```

```
import sequelize from "sequelize";
import database from "../config/database.js";
import Contact from "./contact.js";

const { DataTypes } = sequelize;

const Conversation = database.define("conversation", {
  uuid: {
    type: DataTypes.STRING,
  },
  last_message_created_at: {
    type: DataTypes.DATE,
  },
  last_inbound_message_created_at: {
    type: DataTypes.DATE,
  },
  unread_messages_count: {
    type: DataTypes.INTEGER,
  },
  status: {
    type: DataTypes.INTEGER,
    // values: ['unattended', 'attended', 'inprocess'],
    default: 1,
  },
});

Conversation.belongsTo(Contact, {
    allowNull: false,
    foreignKey: {
      name: "user_contact_id",
    },
    as: "user",
});

export default Conversation;
```

**Comment by EdByrnee** [comment]
I have all my entities in different files but I'm getting the same issue?!

**Comment by TPME** [comment]
Hi,

I have the same issue and it's really blocking me. The weird part is that it seems very inconsistent.

- Models are all in different files
- I use Sequelize({config: {models: ... } }) 

Yet I still have this error. Please help :-/

**Comment by tomfree** [comment]
Have the same problem in my code. Worked with Webpack4 and since upgrding to Webpack 5 this seems to be a problem. To what I figured out so far it seems to be related to circular dependencies due to the Foreign Key relations...

**Comment by nathanlepori** [comment]
The issue still happens when models are in different files but only when target in tsconfig.json is set to any ES version. It does not happen when targeting CommonJS. Maybe @TPME and @EdByrnee can confirm whether they're also targeting ESM. It seems like using the  `() => Model` as a workaround for avoiding circular dependencies only works in CommonJS.

**Comment by schealex** [comment]
I'm having the same issue. Trying the hardest to get it resolved but can't get it working. I'm also using @nestjs/sequelize which does not make it easier because sequelize is encapsulated in there.

Any process or new info on how to resolve this issue?

**Comment by RahmatAliMalik5** [comment]
I was using `@nestjs/sequelize`. Got same error. Updating `module:es2020` and `target:es2015` to `module: commonjs` and `target: es2017` solved the issue.

**Comment by schealex** [comment]
> I was using `@nestjs/sequelize`. Got same error. Updating `module:es2020` and `target:es2015` to `module: commonjs` and `target: es2017` solved the issue.

It does not because changes the target type to commonjs instead of ESM which is not what we want. If you want to build an ESM module that is not a solution!

**Comment by RahmatAliMalik5** [comment]
I am not providing a solution to this issue. I am confirming that it solves when used `commonjs` is used instead of ESM.

**Comment by schealex** [comment]
so there is no future for this project with the now more and more coming ESM projects?

**Comment by nathanlepori** [comment]
**TL;DR**: see workaround code snippet.

Without in-depth knowledge of the code I'm not sure if this could work but I'm suggesting it anyways if someone wants to try to implement it. 

One way of solving the issue might be to allow passing just the model name on one side of the association and keep passing the model getter on the other (basically a TS overload + a `typeof modelGetter === 'string'` check). This way the circular dependency would be avoided (no import required on one side in emitted JS code) and sequelize-typescript could resolve the model constructor from its name from the already registered models.
One problem with this approach is what would happen if the model associating by name is defined before the one associating by getter: in that case sequelize-typescript wouldn't have any constructor to resolve yet... And this isn't really under the user's control since the import order is established by Node.

With that said unfortunately I don't have the time to create a PR myself at the moment but I might do it some time in the future since I will also need this feature. In the meantime maybe @RobinBuschmann has some suggestions on the feasibility of this approach/other possible solutions...

For now though I also suggest a workaround I'm planning on using myself:
It's possible to associate model A with model B normally in A.ts (using the import + decorator). Then leave out the association completely in B.ts and add the association for B in A.ts using regular old sequelize.

Here's an example:

```
// A.ts
import B from './B';

class A extends Model {
  @HasMany(() => B)
  foo!: B[];
}
// Put association in A.ts
B.belongsTo(A);

export default A;

// B.ts
import A from './A';

class B extends Model {
  // Omit association. TS compiler should omit import for A in emitted JS since it's only used as type.
  // @BelongsTo(() => A)
  baz!: A;
}

export default B;
```

It's not super pretty but it should work.
Let me know what do you think/any possible issue with either approach.

**Comment by mschipperheyn** [comment]
What should be happening to avoid this is to call the @ForeignKey, @HasMany etc function with the models object after sequelize model loading so this could function after static initialization and before sync.

I don't buy the whole circular references can't be avoided or don't matter argument. They do matter and they can be avoided.

I always used this snippet before trying the typescript annotation approach, using a static associate function on each model that deals with association for that model. No circular dependencies necessary. 

`
const sequelize = new Sequelize(dbConnectString, options);

const models = loadModels(sequelize); // dynamic model file reading and static initialization 

Object.keys(models).forEach(key => {
        // simplified version
	const model = models[key];
	if (model.associate) {
		model.associate(models);
	}
});

// Model snippet 

static associate = models => {
    Account.hasMany(models.AccountTenant, {
        foreignKey: 'accountId',
        onDelete: 'cascade',
        as: 'roles',
    });
}
`

**Comment by mschipperheyn** [comment]
The root cause of all these issues is the requirement of recursive imports. This PR (https://github.com/RobinBuschmann/sequelize-typescript/pull/1206) solve that in part. But in order to completely solve this, I think the returned reference from an association should be a generic model based on the interface. 

So, the final solution could look something like this:
```

@Table
export class Player extends Model<IPlayer> {

  @Column
  name!: string;

  @Column
  num!: number;

  @ForeignKey(models => models.Team)
  @Column
  teamId!: number;

  @BelongsTo(models => models.Team)
  team!: Model<ITeam>;
}

@Table
export class Team extends Model<ITeam> {

  @Column
  name!: string;

  @HasMany(models => models.Player)
  players!: Model<IPlayer>[];
}
```

**Comment by gugu** [comment]
I'm trying to remove emitDecoratorMetadata option. If you set types/table names manually it can be a solution as well

**Comment by strrife** [comment]
I've got another workaround.

I've noticed that it is possible to reference classes declared later when defining arrays

```
  // this works
  @HasMany(() => AuctionBid)
  public bids: AuctionBid[];

  // this does not work
  @HasOne(() => AuctionBid)
  public lastBid: AuctionBid;
                  ^ ReferenceError: Cannot access 'AuctionBid' before initialization
```

Which brought me to this:

```
  @HasOne(() => AuctionBid)
  public usersBid: ReturnType<() => AuctionBid>;
```

Compiles without errors.

**Comment by remcohaszing** [comment]
Nice find @strrife!

The following alternative works too.

```ts
  @HasOne(() => AuctionBid)
  public usersBid: Awaited<AuctionBid>;
```

**Comment by untimated** [comment]
> I've got another workaround.
> 
> I've noticed that it is possible to reference classes declared later when defining arrays
> 
> ```
>   // this works
>   @HasMany(() => AuctionBid)
>   public bids: AuctionBid[];
> 
>   // this does not work
>   @HasOne(() => AuctionBid)
>   public lastBid: AuctionBid;
>                   ^ ReferenceError: Cannot access 'AuctionBid' before initialization
> ```
> 
> Which brought me to this:
> 
> ```
>   @HasOne(() => AuctionBid)
>   public usersBid: ReturnType<() => AuctionBid>;
> ```
> 
> Compiles without errors.

i did change the Model to Model[] and it works..

**Comment by jose-verissimo** [comment]
In my case changing the order of how you import these models on the same file you use `new Sequelize()` solved the problem, I'd suggest having a look at this repo's example because it does the exact same thing and their version works even though it needs updating to the latest version.

**Comment by d00rsfan** [comment]
I don't believe ESLINT is not screaming: "ESLint: Dependency cycle detected. (import/no-cycle)" on "import" lines on these last workarounds!

**Comment by Matheusleal** [comment]
> In my case changing the order of how you import these models on the same file you use `new Sequelize()` solved the problem, I'd suggest having a look at this repo's example because it does the exact same thing and their version works even though it needs updating to the latest version.

I was getting the same issue, with this change worked fine to me, thanks!

**Comment by Oxicode** [comment]
In case anyone reads this issue, I used forwardRef 
https://docs.nestjs.com/fundamentals/circular-dependency

**Comment by AbandonedCrypt** [comment]
To me the most 'natural' workaround for this is to type related models as union with null.

```ts
  @BelongsTo(() => Model)
  model?: Model | null;
```

as it is logically sound. this will either be included in the query result or not.

**Comment by Sn0wled** [comment]
The solution is explained directly in the readme under the heading "Why () => Model?".

**Comment by henriqueNectaco** [comment]
Please make sure you establish the relationship in only one model. I encountered this error because I was using HasMany in one model and BelongsTo in the other. You only need to use one of these relationships, so just remove the one that's unnecessary.

**Comment by Naktibalda** [comment]
I had this issue today too,
One-way relationships is not a solution, because in many places we want to join the opposite way.
Moving relationships to separate file was possible, but it affected aliases of all foreign keys, so it required huge amount of changes too.

I tried to use both suggestions (with ReturnType and with null), but neither worked, because the problem was decorator using model as value.

So I told both options to Claude AI, and it came up with unexpected solution that actually works for me:

```ts
// At the top of the file
// Import only for type reference
import type { Model as ModelType } from "./model.js";

// Inside the class
@BelongsTo(() => Model)
declare attribute?: ModelType;

// At the bottom of the file
// Import for runtime reference
import { Model } from "./model.js";
```
